### PR TITLE
Refactor as custom resource using aws-sdk

### PR DESCRIPTION
**Pull request by jeffbyrnes**
Wrote this as two commits, in case you wish to publish just the first one as an interim v2.0.0 before the full refactor using `aws-sdk`.

The `aws-s3` gem is long since abandoned, and AWS’ official `aws-sdk` is the way to handle these sorts of operations now.

Advantages:

* Supplying credentials inline is now optional, and if they are omitted,
aws-sdk will follow its now-standard order of operations to find
credentials:
    - Confusingly, the best place to see this order of operations is in the [AWS SDK for Java](http://docs.aws.amazon.com/sdk-for-java/v1/developer-guide/credentials.html)
* We can now specify other regions to use, if desired

### CODE REVIEW

File: resources/default.rb [hunk 1]
```diff
@@ -1,8 +1,34 @@
...
+property :access_key_id,     String
+property :secret_access_key, String
+
+chef_gem 'aws-sdk'
+
+require 'aws-sdk'
+
+Aws.config.update(
+  credentials: Aws::Credentials.new(access_key_id, secret_access_key)
+) if access_key_id && secret_access_key
```
Comment (eherot) [inline resources/default.rb]: Will this fall back to role-based permissions without requiring any additional parameters?

File: providers/default.rb [hunk 2]
```diff
@@ -1,32 +0,0 @@
-action :upload do
-  fail 'remote_path required for upload' if new_resource.remote_path.empty?
```
Comment (eherot) [inline providers/default.rb]: It looks like we lost this check...

File: metadata.rb [hunk 3]
```diff
@@ -7,3 +7,7 @@
 version          '1.0.2'
 
 supports 'ubuntu', '>= 12.04'
+
+chef_version '~> 12.1'
+
+depends 'compat_resource', '~> 12.16'
```
Comment (eherot) [inline metadata.rb]: Can we use `>=` for `depends` that are this widely used? These pessimistic dependencies cause lots of unnecessary headaches.

File: resources/default.rb (hunk 1 above)
Comment (jeffbyrnes) [inline resources/default.rb]: It sure will! The link in the description lays out the precedence nicely, though of course, mentally substitute “Ruby config options” where it talks about “Java properties”.

File: providers/default.rb (hunk 2 above)
Comment (jeffbyrnes) [inline providers/default.rb]: Unnecessary; it’s a required property (see the `property :remote_path` definition above).

File: metadata.rb (hunk 3 above)
Comment (jeffbyrnes) [inline metadata.rb]: The `compat_resource` cookbook tracks Chef’s own version, so that may be unwise, since Chef 13 is planned to be quite breaking.

My usual path for upgrading a cookbook like this is to use these version constraints in a breaking release, then issue _another_ breaking release, dropping `compat_resource`, and specifying `chef_version '>= 12.5'`.

All that said, nobody but you and (soon) me are using this cookbook right now, so I’ll just go ahead & skip to that part.

File: providers/default.rb (hunk 2 above)
Comment (eherot) [inline providers/default.rb]:
```ruby
[1] pry(main)> remote_path = ''
=> ""
[2] pry(main)> puts "It's set" if remote_path
It's set
=> nil
[3] pry(main)> puts "But it's still empty" if remote_path.empty?
But it's still empty
=> nil
```

Comment (eherot) [inline providers/default.rb]: Also, it's [always been a required property](https://github.com/evertrue/s3_put-cookbook/pull/3/files#diff-5c4d569da2ded311729109c9ef7c0641L6)

Comment (jeffbyrnes) [inline providers/default.rb]: Then that would mean this check has always been irrelevant. If a required property is missing from a resource, and lacks a default value for that property, Chef will throw an error.

Perhaps the behavior was different or looser back in the days before Custom Resources?

Comment (eherot) [inline providers/default.rb]: I think you're misunderstanding my example. If someone were to do this with the current code, it would fail:
```ruby
s3_put '/path/to/my_file` do
  remote_file ''
end
```
...but with your code it would not.

File: resources/default.rb (hunk 1 above)
Comment (eherot) [inline resources/default.rb]: 🎉

File: providers/default.rb (hunk 2 above)
Comment (jeffbyrnes) [inline providers/default.rb]: That’s true, but that _is_ an empty value: it corresponds to the root of the bucket. Granted, that is probably better represented as `remote_file '/'`, but the resource corrects for that (speaking of which, I just rebased the last commit to do that consistently across both actions).

If you still feel very strongly that an empty string is an illegal value, I’ll add a validator in to the property.

**Comment by jeffbyrnes** [comment]
@eherot by the way, if you merge in #1, I’ll rebase this branch to run those tests.

File: providers/default.rb (hunk 2 above)
Comment (eherot) [inline providers/default.rb]: That is a behavior change (though not necessarily a bad one) so it should probably at the very least be a different PR.

Comment (jeffbyrnes) [inline providers/default.rb]: I’m good with that.

I figured that, since this entire PR is breaking, owing to the adoption of `aws-sdk` instead of `aws-s3`, it was alright to roll it in. I’ll break it out into its own change and we can discuss its merits separately.

File: resources/default.rb [hunk 4]
```diff
@@ -1,8 +1,34 @@
...
+
+s3 = Aws::S3::Resource.new region: region
+
+remote_path << '/' unless remote_path.end_with? '/'
+
+action :upload do
+  s3.bucket(bucket).object(remote_path + filename).upload source_file
```
Comment (eherot) [inline resources/default.rb]: Also:
```ruby
> s3.bucket(bucket).object(remote_path + filename).upload source_file
NoMethodError: undefined method `upload' for #<Aws::S3::Object:0x007fa7a9ac3980>
```

Comment (eherot) [inline resources/default.rb]: I think you meant to use `upload_file`

Comment (jeffbyrnes) [inline resources/default.rb]: Huh. I guess the methods are chainable. I figured they would be, based on the [uploading an item code example](http://docs.aws.amazon.com/sdk-for-ruby/v2/developer-guide/s3-example-upload-bucket-item.html) that AWS provides.

Comment (jeffbyrnes) [inline resources/default.rb]: Oh dammit, I mistyped that method name

Comment (jeffbyrnes) [inline resources/default.rb]: Ok, fixed it, that is now `upload_file`.

**Comment by jeffbyrnes** [comment]
@eherot seems like there should be some tests. I’ll have to ponder on how to do that, since I’m not sure there’s a built-in way to mock.

**Comment by eherot** [comment]
I wonder if you could do something with [this](http://docs.aws.amazon.com/sdkforruby/api/Aws/ClientStubs.html)

**Comment by jeffbyrnes** [comment]
@eherot at the moment, leaving aside tests, whatcha think of this?
